velox/functions/sparksql/Split.h

```diff
@@ -0,0 +1,168 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/functions/lib/Utf8Utils.h"
+
+namespace facebook::velox::functions::sparksql {
+
+/// split(string, delimiter[, limit]) -> array(varchar)
+///
+/// Splits string on delimiter and returns an array of size at most limit.
+/// delimiter is a string representing regular expression.
+/// limit is an integer which controls the number of times the regex is applied.
+/// By default, limit is -1, which means 'no limit', the delimiter will be
+/// applied as many times as possible.
+template <typename T>
+struct Split {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // Results refer to strings in the first argument.
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Array<Varchar>>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& delimiter) {
+    doCall(result, input, delimiter, INT32_MAX);
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Array<Varchar>>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& delimiter,
+      const arg_type<int32_t>& limit) {
+    doCall(result, input, delimiter, limit > 0 ? limit : INT32_MAX);
+  }
+
+ private:
+  void doCall(
+      out_type<Array<Varchar>>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& delimiter,
+      int32_t limit) const {
+    if (delimiter.empty()) {
+      splitEmptyDelimiter(result, input, limit);
+    } else {
+      split(result, input, delimiter, limit);
+    }
+  }
+
+  // When pattern is empty, split each character out. Since Spark 3.4, when
+  // delimiter is empty, the result does not include an empty tail string, e.g.
+  // split('abc', '') outputs ["a", "b", "c"] instead of ["a", "b", "c", ""].
+  // The result does not include remaining string when limit is smaller than the
+  // string size, e.g. split('abc', '', 2) outputs ["a", "b"] instead of ["a",
+  // "bc"].
+  void splitEmptyDelimiter(
+      out_type<Array<Varchar>>& result,
+      const arg_type<Varchar>& input,
+      int32_t limit) const {
+    if (input.size() == 0) {
+      result.add_item().setNoCopy(StringView());
+      return;
+    }
+
+    const size_t end = input.size();
+    const char* start = input.data();
+    size_t pos = 0;
+    int32_t count = 0;
+    while (pos < end && count < limit) {
+      auto charLength = tryGetCharLength(start + pos, end - pos);
+      if (charLength <= 0) {
+        // Invalid UTF-8 character is treated as single character.
+        charLength = 1;
```

Comment:
Given the following behavior of tryGetCharLength, would it make sense to set charLength to -charLength ? If not, why? 

> A negative return value means
> /// the UTF-8 sequence at the position is invalid, and the length of the invalid
> /// sequence is the absolute value of the result.

Reply:
thanks @mbasmanova , i didn't before noticed this comment, but it's cool if the absolute value of the return value is the invalid character sequence length, then we can levearge that, instead of processing the invalid character sequence one by one.

After that, the continuous invalid character sequence for example `\xED\xA0` would be treat one not two `\xED`, `\xA0`.